spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/neo4j/Neo4jImperativeDataConfiguration.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.data.neo4j;
+
+import org.neo4j.driver.Driver;
+
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.data.ConditionalOnRepositoryType;
+import org.springframework.boot.autoconfigure.data.RepositoryType;
+import org.springframework.boot.autoconfigure.neo4j.Neo4jDriverAutoConfiguration;
+import org.springframework.boot.autoconfigure.transaction.TransactionManagerCustomizers;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.core.annotation.Order;
+import org.springframework.data.neo4j.config.Neo4jDefaultCallbacksRegistrar;
+import org.springframework.data.neo4j.core.DatabaseSelectionProvider;
+import org.springframework.data.neo4j.core.Neo4jClient;
+import org.springframework.data.neo4j.core.Neo4jOperations;
+import org.springframework.data.neo4j.core.Neo4jTemplate;
+import org.springframework.data.neo4j.core.mapping.Neo4jMappingContext;
+import org.springframework.data.neo4j.core.transaction.Neo4jTransactionManager;
+import org.springframework.data.neo4j.repository.config.Neo4jRepositoryConfigurationExtension;
+import org.springframework.transaction.PlatformTransactionManager;
+
+/**
+ * Internal configuration of Neo4j client and transaction manager.
+ *
+ * @author Michael J. Simons
+ */
+@Configuration(proxyBeanMethods = false)
+@ConditionalOnClass({ Neo4jTransactionManager.class, PlatformTransactionManager.class })
+@ConditionalOnRepositoryType(store = "neo4j", type = RepositoryType.IMPERATIVE)
+@AutoConfigureAfter(Neo4jDriverAutoConfiguration.class)
+@AutoConfigureBefore(Neo4jImperativeRepositoriesConfiguration.class)
+@Import(Neo4jDefaultCallbacksRegistrar.class)
+final class Neo4jImperativeDataConfiguration {
+
+	@Bean("databaseSelectionProvider")
+	@ConditionalOnProperty(prefix = "spring.data.neo4j", name = "database")
+	@ConditionalOnMissingBean
+	@Order(-30)
```

Comment:
Can you share why this is required?

Reply:
It should have more importance than the `defaultSelectionProvider` underneath this method to ensure that the configured one has precedence.

Reply:
Thanks for the feedback. We don't use `@Order` that way anywhere else so I think we should express this order in a different manner. Usually we move the bits to two separate `@Configuration` that we import in order. Look, for instance, at `org.springframework.boot.autoconfigure.r2dbc.ConnectionFactoryConfigurations`.

Reply:
On a second review, I think you should do this check programmatically within one single `@Bean` method, checking for the property using the `Environment`.